packages/bpk-component-popover/src/BpkPopover.tsx

```diff
@@ -0,0 +1,240 @@
+
+// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
+import BpkCloseButton from '../../bpk-component-close-button';
+// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
+import { BpkButtonLink } from '../../bpk-component-link';
+import BpkText, { TEXT_STYLES } from '../../bpk-component-text';
+import { TransitionInitialMount, cssModules } from '../../bpk-react-utils';
+
+import { ARROW_ID } from './constants';
+
+import STYLES from './BpkPopover.module.scss';
+
+const getClassName = cssModules(STYLES);
+
+const EVENT_SOURCES = {
+  CLOSE_BUTTON: 'CLOSE_BUTTON',
+  CLOSE_LINK: 'CLOSE_LINK',
+};
+
+const bindEventSource =
+  (source: string, callback: any, event: SyntheticEvent<HTMLButtonElement>) => {
+    if (event.persist) {
+      event.persist();
+    }
+
+    callback(event, { source });
+  };
+
+export type Props = {
+  children: ReactNode;
+  closeButtonText: string;
+  id: string;
+  label: string;
+  onClose: (
+    event: SyntheticEvent<HTMLButtonElement>,
+    props: { source: (typeof EVENT_SOURCES)[keyof typeof EVENT_SOURCES] },
+  ) => void;
+  className?: string | null;
+  closeButtonIcon?: boolean;
+  closeButtonProps?: Object;
+  isOpen?: boolean;
+  labelAsTitle?: boolean;
+  padded?: boolean;
+  placement?: 'top' | 'right' | 'bottom' | 'left';
+  showArrow?: Boolean;
+  target: ReactElement<any>;
+};
+
+const BpkPopover = ({
+  children,
+  className = null,
+  closeButtonIcon = true,
+  closeButtonProps = {},
+  closeButtonText,
+  id,
+  isOpen = false,
+  label,
+  labelAsTitle = false,
+  onClose,
+  padded = true,
+  placement = 'bottom',
+  showArrow = true,
+  target,
+  ...rest
+}: Props) => {
+  const [isOpenState, setIsOpenState] = useState(isOpen);
+  const arrowRef = useRef(null);
+
+  const { context, floatingStyles, refs } = useFloating({
+    open: isOpenState,
+    onOpenChange: setIsOpenState,
+    placement,
+    middleware: [
+      showArrow && offset(17),
+      flip({ crossAxis: true }),
+      shift(),
+      showArrow && arrow({ element: arrowRef }),
+    ],
+    whileElementsMounted: autoUpdate,
+  });
+
+  const click = useClick(context);
+  const dismiss = useDismiss(context);
+
+  // Merge all the interactions into prop getters
+  const { getFloatingProps, getReferenceProps } = useInteractions([
+    click,
+    dismiss,
+  ]);
+
+  const targetElement = isValidElement(target) ? (
+    cloneElement(target, {
+      ...getReferenceProps(),
+      // @ts-ignore - we're adding a popover ref to the target element so we can position the popover relative to it
+      ref: refs.setReference,
+    })
+  ) : (
+    <div ref={refs.setReference} {...getReferenceProps()}>
+      {target}
+    </div>
+  );
+  const classNames = getClassName('bpk-popover', className);
+  const bodyClassNames = getClassName(padded && 'bpk-popover__body--padded');
+
+  const labelId = `bpk-popover-label-${id}`;
+
+  return (
+    <>
+      {targetElement}
```

Comment:
Why do _we_ need to render the targetElement rather than letting the consumer do it as it was before?

Reply:
And related to this - I think it'd be good to create a short migration guide for consumers?

Reply:
> think it'd be good to create a short migration guide for consumers?

Yep on the todo!

> Why do we need to render the targetElement rather than letting the consumer do it as it was before?

This is because we need to attach the generated refs from the floating UI library to it for all its magic and handling of it. So if we don't provide it there is no where for the refs to be provided and so the popover won't attach and operate.

Doing digging through our code folks are only ever passing the target in anyway instead of rendering separately so this is no change to what we were doing before and what was required of the consumer :) You can see this in the now deleted PopoverPortal here: https://github.com/Skyscanner/backpack/pull/3390/files#diff-03fea3ca881f4c6ce5e50658360c673e40526260b7fb538ffdd5ee36c51d3af9L197

Reply:
Awesome, makes sense!